lib/worker/output.go

```diff
@@ -0,0 +1,219 @@
+	"io"
+	"sync"
+	"time"
+)
+
+// OutputHandler manages output buffering and forwarding to
+// concurrent output listeners.
+//
+// It is initialised with one or more readers which it
+// reads to buffer the content and send to listeners
+//
+// OutputHandler attempts to forward output by chunks of
+// length 76.
+// TODO: the chunk size should be configurable
+//
+// Forwarded content for a new listener is always sent from
+// the start. Updates are always sent in the order they were
+// read. This means output is always in order by reader but
+// not necessarily across readers.
+//
+// TODO: This could possibly be an interface so that the way
+// we forward output is configurable. For example, we might
+// have an output handler that sends output based on time
+// ellapsed.
+//
+// TODO: At the moment, all OutputHandlers are kept in memory.
+// This could be bad if we are running many commands with
+// large output (which is not uncommon at all). Consider
+// storing into files after a certain period has passed
+// or using a database.
+type OutputHandler struct {
+	listeners []*Listener
+	// readers is the readers that the output handler is extracting output from
+	readers []io.Reader
+	// combinedBuffer keeps the combined output, this is used to forward to
+	// new listeners that came in late
+	combinedBuffer []byte
+	// messages is a channel where new output is delivered
+	messages chan []byte
+	// errors is a channel to forward output reading errors to listeners
+	errors chan error
+	// done is a channel to notify listeners that all output has been read
+	done chan struct{}
+	mu   *sync.Mutex
+}
+
+// ProcessOutputEntry is the struct sent to clients' channels
+// streaming output
+type ProcessOutputEntry struct {
+	Content    []byte
+	ReceivedAt time.Time
+}
+
+// Listener represents an output listener that the output handler will
+// send updates to.
+type Listener struct {
+	outputChan chan ProcessOutputEntry
+	errorChan  chan error
+}
+
+// NewOutputHandler returns a new OutputHandler struct from the provided readers.
+func NewOutputHandler(rc ...io.Reader) (*OutputHandler, error) {
+	if len(rc) == 0 {
+		return nil, fmt.Errorf("must provide at least one io.Reader")
+	}
+	o := &OutputHandler{
+		readers:  rc,
+		mu:       &sync.Mutex{},
+		done:     make(chan struct{}),
+		messages: make(chan []byte),
+		errors:   make(chan error),
+	}
+	go o.handleOutput()
+	go o.handleBroadcast()
+	go o.handleErrors()
+	return o, nil
+}
+
+// Stream returns a channel to receive output up to the
+// present time and stream real time. The channel closes when
+// there the process the output is coming from is finished.
+// Any errors encountered during output reading are sent through
+// the errors channel.
+func (o *OutputHandler) Stream() (chan ProcessOutputEntry, chan error) {
+	output := make(chan ProcessOutputEntry)
+	errChan := make(chan error)
+
+	go func() {
+		if o.bufferLen() > 0 {
+			output <- ProcessOutputEntry{Content: o.combinedBuffer}
```

Comment:
I think I might need some more comments around this. Is this implementing the catchup part of the streamed output?

Reply:
Yep, this is what this is. The combined buffer is a []byte that contains all output up to the "present".  When a new listener comes in, we send the whole slice through unless it is empty. Added a comment to clarify.